File: s3-inf-ppo/laboratoria/lab04/lab04.php
<?php

class WinnerWasCalled extends Exception {}

class Log {

	static function info(string $message = ""): void {
		echo $message . PHP_EOL;
	}

}

class Dice {

	function roll(): int {
		$result = rand(1, 6);
		Log::info("Dice roll: " . $result);
		return $result;
	}

}

class Pawn {

	public $position;
	public $name;

	function __construct(string $name) {
		$this->position = 0;
		$this->name = $name;
		
		Log::info("$name joined the game.");
	}

}

class Board {

	const MAX_POSITION = 100;

	public $pawns;
	public $dice;
	public $winner;
	public $turnsCounter;

	function __construct() {
		$this->pawns = [];
		$this->dice = null;
		$this->winner = null;
		$this->turnsCounter = 0;
	}

	function performTurn(): void {
		$this->turnsCounter++;
		Log::info();
		Log::info("Turn " . $this->turnsCounter);

		foreach($this->pawns as $pawn) {
			$rollResult = $this->dice->roll();
			$pawn->position += $rollResult;
			Log::info($pawn->name . " new position: " . $pawn->position);

			if($pawn->position >= self::MAX_POSITION) {
				$this->winner = $pawn;
				throw new WinnerWasCalled();
			}
		}
	}

}

$board = new Board();	
$board->dice = new Dice();
$board->pawns[] = new Pawn("Luke Skywalker");
$board->pawns[] = new Pawn("Darth Vader");

try {
	while(true) {
		$board->performTurn();
	}
} catch(WinnerWasCalled $exception) {
	Log::info();
	Log::info($board->winner->name . " won.");
}

File: s3-inf-ppo/laboratoria/lab04/cs/lab04.cs
using System;
using System.Collections.Generic;

public class WinnerWasCalled : Exception { }

public class Log
{
    public static void Info(string message = "")
    {
        Console.WriteLine(message);
    }
}

public class Dice
{
    public Random rand = new Random();

    public int Roll()
    {
        int result = (rand.Next(6) + 1);
        Log.Info("Dice roll: " + result);

        return result;
    }
}

public class Pawn
{
    public int position;
    public string name;

    public Pawn(string name)
    {
        this.name = name;
        this.position = 0;

        Log.Info(name + " joined the game.");
    }
}

public class Board
{
    const int maxPosition = 100;

    public List<Pawn> pawns;
    public Dice dice;
    public Pawn winner;
    public int turnsCounter;

    public Board()
    {
        this.turnsCounter = 0;
        this.pawns = new List<Pawn>();
    }

    public void PerformTurn()
    {
        this.turnsCounter++;
        Log.Info();
        Log.Info("Turn " + this.turnsCounter);

        for (int i = 0; i < this.pawns.Count; i++)
        {
            int rollResult = this.dice.Roll();
            Pawn pawn = this.pawns[i];
            pawn.position += rollResult;

            Log.Info(pawn.name + " new position: " + pawn.position);

            if (pawn.position >= maxPosition)
            {
                this.winner = pawn;
                throw new WinnerWasCalled();
            }
        }
    }
}

public class Lab04
{
    public static void Main(string[] args)
    {
        Board board = new Board();
        board.dice = new Dice();
        board.pawns.Add(new Pawn("Luke Skywalker"));
        board.pawns.Add(new Pawn("Darth Vader"));

        try
        {
            while (true)
            {
                board.PerformTurn();
            }
        }
        catch (WinnerWasCalled ex)
        {
            Log.Info();
            Log.Info(board.winner.name + " won");
            Console.ReadKey();
        }
    }
}


